include/fi_shm.h

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2016-2017 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_SHM_H_
+#define _FI_SHM_H_
+
+#include "config.h"
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include <fi_atom.h>
+#include <fi_proto.h>
+#include <fi_mem.h>
+#include <fi_rbuf.h>
+
+#include <rdma/providers/fi_prov.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+#define SMR_VERSION	1
+
+#ifdef HAVE_ATOMICS
+#define SMR_FLAG_ATOMIC	(1 << 0)
+#else
+#define SMR_FLAG_ATOMIC	(0 << 0)
+#endif
+
+#if ENABLE_DEBUG
+#define SMR_FLAG_DEBUG	(1 << 1)
+#else
+#define SMR_FLAG_DEBUG	(0 << 1)
+#endif
+
+
+/* SMR op_data: Specifies data source location */
+enum {
+	smr_op_inline,	/* command data */
+	smr_op_inject,	/* inject buffers */
+	smr_op_iov,	/* reference iovec via CMA */
+};
+
+struct smr_cmd_hdr {
+	struct ofi_op_hdr	op;
+	uint32_t		cmd_id;
+	uint32_t		rx_key;
+};
+
+#define SMR_CMD_SIZE		128	/* align with 64-byte cache line */
+#define SMR_CMD_DATA_LEN	(128 - sizeof(struct smr_cmd_hdr))
+
+#define SMR_NAME_SIZE	32
+struct smr_addr {
+	char	name[SMR_NAME_SIZE];
+};
+
+union smr_cmd_data {
+	uint8_t			msg[SMR_CMD_DATA_LEN];
+	struct iovec		iov[SMR_CMD_DATA_LEN / sizeof(struct iovec)];
+	struct ofi_rma_iov	rma_iov[SMR_CMD_DATA_LEN / sizeof(struct ofi_rma_iov)];
+	struct ofi_rma_ioc	rma_ioc[SMR_CMD_DATA_LEN / sizeof(struct ofi_rma_ioc)];
+};
+
+struct smr_cmd {
+	struct smr_cmd_hdr	hdr;
+	union smr_cmd_data	data;
+};
+
+enum {
+	SMR_INJECT_SIZE = 4096
+};
+
+struct smr_region;
+DECLARE_FREESTACK(struct smr_region *, smr_peer);
+
+#define SMR_MAX_PEERS	256
+
+struct smr_map {
+	fastlock_t	lock;
```

Comment:
Depends what the use is. The map lock is used when adding peer regions to the map so if you have different threads/endpoints sharing the same map, then we would need a per map lock since the map wouldn't be associated with a specific region to lock. If that's not a case intended to be supported, then we could just use the smr_region lock. My gut says keep them separate, though.

Reply:
trust your gut --  that sounds like a slogan for a bad pizza joint